### DISCUSSION TITLE
Custom Sounds attribute

### DISCUSSION THREAD

**Discussion by JessicaTheGunLady** [nosoop/SM-TFCustAttr#4] [Created attributes]
[This plugin](https://github.com/JessicaTheGunLady/SM-TF2-ReplaceSound) adds the ability to replace weapon sounds with custom ones, primarily intended for the CWX plugin which up until now hasn't had a way of doing this which I'm publically aware of, it works by having a `sound index` attribute, then checking a KeyValues file for that sound index, and then looking inside there for sounds it needs to replace.

As it is I'm not even sure if I *want* to post this here since I know it's absolutely flawed code that could do with a lot of improving, it's been the product of 5 days of constant work with my brain half the time being tired or cloudy, but it at least seems to resemble a functional state and I'd rather make sure that people are aware that this plugin exists in case it can legitimately be of use, or if people want to improve it to a point where it *can* be of use, some improvements I have tried to make but ultimately refused to work likely due to my own mistakes, so this is a compromise between having a well-written plugin in 5 years and a horrendously optimised and more limited plugin 4 days ago.

**Comment by JessicaTheGunLady** [comment]
Turns out this plugin actually wasn't quite functional, only working for a singular custom weapon at any given time, _probably_ due to a mistake... Or several... That I made, thankfully having it out there has given my brain some breathing room and I'll probably be able to figure out the problems.

**Comment by nosoop** [comment]
Thanks for posting!  There isn't any implementation that I'm aware of either.

I've had a few ideas for one but they required some functionality that I haven't gotten around to implementing within CWX (specifically, a native function to obtain `KeyValues` structures by name from CWX item definitions).  ~~Because I'm lazy~~

Having an external data file does make your plugin more flexible as well; it should be usable outside of CWX.

At a glance it looks like you're parsing and passing around `KeyValues` via files; having to touch files is very slow.  You'll want to have a `KeyValues` variable in global scope and ideally should only read from the file once.

**Reply by JessicaTheGunLady** [reply]
Good news is I've actually since been working on a new iteration which works a bit differently, new data isn't getting stored to a database, though this plugin is actually using the hidden dev attributes method of implementing attributes instead, that way I can actually check during the SoundHook function and not kill people's servers, though that does mean that this plugin is no longer being used which does truthfully feel a bit saddening to say, but hey, the plugin's at least _mostly_ functional?

I appreciate the word on KeyValues though, I'll certainly be trying to keep it in mind in the future at the very least.

**Reply by nosoop** [reply]
Attribute injection in general is great, even if I've run into a number of quirky edge cases with the hidden dev attributes version.

In case you were curious, [here's the set of microbenchmarks I've done between native and CAF-based attributes](https://gist.github.com/nosoop/e19dd9c103bdc1c697a3e83603bf7725).
